### PR TITLE
feat: support cohort targeting for local evaluation 

### PR DESCRIPTION
**Pull request by tyiuhc**
### Summary

Support cohort targeting for local evaluation 

### Checklist

* [X] Does your PR title have the correct [title format](https://github.com/amplitude/experiments-ruby-server/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:   No

### CODE REVIEW

File: lib/experiment/cohort/cohort_loader.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+          ensure
+            remove_job(cohort_id)
+          end
+          @jobs[cohort_id] = future
+        end
+        @jobs[cohort_id]
+      end
+    end
+
+    def update_stored_cohorts
```
Comment (bgiori) [inline lib/experiment/cohort/cohort_loader.rb:29]: Can we remove this function and just call load_cohort for each cohort stored in storage instead?

File: lib/experiment/cohort/cohort_sync_config.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+  # Experiment Cohort Sync Configuration
+  class CohortSyncConfig
+    # This configuration is used to set up the cohort loader. The cohort loader is responsible for
+    # downloading cohorts from the server and storing them locally.
+    #   Parameters:
+    #     api_key (str): The project API Key
+    #     secret_key (str): The project Secret Key
+    #     max_cohort_size (int): The maximum cohort size that can be downloaded
+    #     cohort_request_delay_millis (int): The delay in milliseconds between cohort download requests
```
Comment (bgiori) [inline lib/experiment/cohort/cohort_sync_config.rb:13]: I would specify that this delay is only applied on failure.

File: lib/experiment/cohort/cohort_loader.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+          rescue StandardError => e
+            [cohort_id, e] # Return the cohort_id and the error
+          end
+        end
+
+        results = Concurrent::Promises.zip(*futures).value!
+
+        # Collect errors from the results
+        results.each do |result|
+          errors << result if result.is_a?(Array) && result[1].is_a?(StandardError)
```
Comment (bgiori) [inline lib/experiment/cohort/cohort_loader.rb:45]: `[1]`?

File: lib/experiment/deployment/deployment_runner.rb [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    def update_flag_configs
+      flags = @flag_config_fetcher.fetch_v2
+      flag_configs = flags.each_with_object({}) { |flag, hash| hash[flag['key']] = flag }
+      flag_keys = flag_configs.values.map { |flag| flag['key'] }.to_set
+      @flag_config_storage.remove_if { |f| !flag_keys.include?(f['key']) }
+
+      unless @cohort_loader
+        flag_configs.each do |flag_config|
+          flag_config = flag_config[1]
```
Comment (bgiori) [inline lib/experiment/deployment/deployment_runner.rb:64]: What does the `[1]` do?

File: lib/experiment/deployment/deployment_runner.rb [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        new_cohort_ids.merge(AmplitudeExperiment.get_all_cohort_ids_from_flag(flag_config))
+      end
+
+      existing_cohort_ids = @cohort_storage.cohort_ids
+      cohort_ids_to_download = new_cohort_ids - existing_cohort_ids
+
+      futures = cohort_ids_to_download.map do |cohort_id|
+        future = @cohort_loader.load_cohort(cohort_id)
+        future.on_rejection do |reason|
+          @logger.warn("Download cohort #{cohort_id} failed: #{reason}")
```
Comment (bgiori) [inline lib/experiment/deployment/deployment_runner.rb:83]: probably want to do an error log

**Review by bgiori — CHANGES_REQUESTED** [review]
We should make it so that any use of the cohort loader goes through the load_cohort function.

**Review by bgiori — APPROVED** [review]

**Review by zhukaihan — APPROVED** [review]

**Pull request merged**